src/Gasworks.sol

```diff
@@ -49,39 +54,105 @@ contract Gasworks is ERC2771Recipient, Owned {
     }
 
     function setTokens(address _token) external onlyOwner {
-        tokens[_token]=true;
+        tokens[_token] = true;
     }
 
     function isPermitted(address _token) public view returns (bool) {
         return tokens[_token];
     }
 
-    function swapWithPermit(PermitData calldata permit, SwapData calldata swapData) external {
+    /**
+     * Swaps an exact amount of SetTokens in 0x for a given amount of ERC20 tokens.
+     * Using a safePermit for the ERC20 token transfer
+     * Method must be called by BiconomyForwarder
+     *
+     * @param permit              Permit data of the ERC20 token used (USDC)
+     * @param swapData            Data of the swap to perform
+     */
+    function swapWithPermit(PermitData calldata permit, SwapData calldata swapData) external isBiconomy {
         require(isPermitted(permit._tokenContract), "INVALID_SELL_TOKEN");
         require(isPermitted(swapData.buyToken), "INVALID_BUY_TOKEN");
-        // Check if Biconomy is sender
 
         ERC20 token = ERC20(permit._tokenContract);
         safePermit(token, permit);
 
         token.safeTransferFrom(permit._owner, address(this), permit._amount);
 
-        emit Received(permit._owner, permit._tokenContract, permit._amount);
+        emit Received(permit._owner, permit._tokenContract, permit._amount, msg.sender);
 
         _fillQuoteInternal(swapData, permit._amount, permit._owner, permit._tokenContract);
     }
 
-    function _fillQuoteInternal(SwapData calldata swap, uint256 sellAmount, address _owner, address _sellToken) internal {
+    /**
+     * Issues an exact amount of SetTokens for given amount of input ERC20 tokens.
+     * Using a safePermit for the ERC20 token transfer
+     * The excess amount of tokens is returned in an equivalent amount of ether.
+     * Method must be called by BiconomyForwarder
+     *
+     * @param permit              Permit data of the ERC20 token used (USDC)
+     * @param mintData            Data of the issuance to perform
+     */
+    function mintWithPermit(PermitData calldata permit, MintData calldata mintData) external isBiconomy {
+        require(isPermitted(permit._tokenContract), "INVALID_SELL_TOKEN");
+        require(isPermitted(address(mintData._setToken)), "INVALID_TOKEN_TO_MINT");
+
+        ERC20 token = ERC20(permit._tokenContract);
+        safePermit(token, permit);
+
+        token.safeTransferFrom(permit._owner, address(this), permit._amount);
+
+        emit Received(permit._owner, permit._tokenContract, permit._amount, msg.sender);
+
+        token.safeApprove(address(exchangeIssuance), mintData._maxAmountInputToken);
+
+        exchangeIssuance.issueExactSetFromToken(
+            mintData._setToken,
+            IERC20(permit._tokenContract),
+            mintData._amountSetToken,
+            mintData._maxAmountInputToken,
+            mintData._componentQuotes,
+            mintData._issuanceModule,
+            mintData._isDebtIssuance
+        );
+
```

Comment:
[question] isn't it a good idea to verify that the amount of the token minted is at least what is required to be issue in the mintData struct?

Reply:
That validation is already done by the exchangeIssuance contract inside the issueExactSetFromToken function, altho the function would still fail by the next line anyway if the amount is not the required